docs(community-wallets): add developer guide for community wallets (#1040)
midnightntwrk/midnight-docs: sdks/community/wallets/cli-and-mcp.mdx

---
SPDX-License-Identifier: Apache-2.0
copyright: This file is part of midnight-docs. Copyright (C) Midnight Foundation. Licensed under the Apache License, Version 2.0 (the "License"); You may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0 Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.
sidebar_position: 30
sidebar_label: CLI and MCP
description: "Develop and test on the Midnight Network without a browser wallet using the wallet CLI, and drive a self-custody wallet from AI agents over MCP."
title: Develop with the wallet CLI and MCP
tags: [community, wallets, cli, mcp, ai-agents, testing, automation, devnet]
toc_max_heading_level: 2
id: community-wallets-cli-mcp
---

# Develop with the wallet CLI and MCP

> The **CLI** and **MCP** columns of the [custody and interface model](./wallets-overview.mdx#the-custody-and-interface-model): how scripts, CI, and AI agents hold and move funds on Midnight without a browser extension.

Reviewer: [MAJOR] Same issue as integration.mdx:15. "The CLI and MCP columns of the custody and interface model" assumes the reader has seen the overview. Open with what the page covers:

"How scripts, CI, and AI agents hold and move funds on Midnight without a browser extension, using the wallet CLI and its MCP server."

Author: Done — opens with what the page covers (scripts, CI, and agents holding/moving funds via the wallet CLI and its MCP server), then links back to the overview. (04141105)

midnight dust register                         # REQUIRED before transfers (else error 138)
midnight balance
midnight transfer mn_addr_undeployed1... 100   # positional: <recipient> <amount>

midnight serve --network preview               # headless DApp Connector on ws://localhost:9932
midnight serve --network preview --approve-all # dev only: auto-approve every request
```

:::warning[#1 beginner gotcha]

`dust register` **before** transferring, or it fails with `BalanceCheckOverspend` (138). DUST regenerates about 5 min locally (see [Fees and DUST](./integration.mdx#fees-and-dust), [Troubleshooting](./reference.mdx#troubleshooting-and-error-reference)).

:::

```ts
// A DApp connects to the CLI connector exactly like a browser wallet:
import { createWalletClient } from "midnight-wallet-connector";
const wallet = await createWalletClient({ url: "ws://localhost:9932", networkId: "Preview" });
const balances = await wallet.getUnshieldedBalances(); // same ConnectedAPI surface
```

Build and test against `midnight serve` (fast, scriptable, no pop-ups), then deploy and let users connect with Lace/1AM, zero code changes.

:::note[Freshness (v0.4.1)]

> `midnight airdrop` can now **fund any address directly** (paste the recipient's address, no seed import), which removes a real annoyance when funding browser wallets on a local network. The indexer connection also moved to the **v4 API**, fixing connection errors people hit on the test networks. ([repo](https://github.com/nel349/midnight-wallet-cli))

Reviewer: "removes a real annoyance" is informal/editorial.

```suggestion
`midnight airdrop` can now **fund any address directly** (paste the recipient's address, no seed import needed), so you no longer need to import a seed to fund browser wallets on a local network. The indexer connection also moved to the **v4 API**, fixing connection errors on the test networks. ([repo](https://github.com/nel349/midnight-wallet-cli))
```

Author: Done — applied the neutral phrasing. (04141105)

:::

([CLI README](https://www.npmjs.com/package/midnight-wallet-cli), [connector](https://www.npmjs.com/package/midnight-wallet-connector), [local playground](https://forum.midnight.network/t/local-playground-for-midnight-compact-contracts-run-a-full-node-indexer-and-proof-server-via-docker-fund-your-lace-wallet-and-deploy-without-testnets-or-faucets/1002))

## AI and agent tooling (the MCP column)

The **MCP** column, *agents calling tools*, in two flavors.

> **Wallet-control MCP.** The **midnight-wallet-cli** ships an **MCP server** (`midnight-wallet-mcp`) with **31 tools** so an AI agent can operate a self-custody wallet, the **Self-custody × MCP** cell. The tools mirror the CLI: **wallet and info** (`midnight_wallet_generate/_list/_use/_info/_remove`, `midnight_info`, `midnight_address`, `midnight_genesis_address`), **transactions** (`midnight_balance`, `_transfer`, `_airdrop`, `midnight_inspect_cost`), **DUST** (`midnight_dust_register/_status`), **local network** (`midnight_localnet_up/_stop/_down/_status/_clean`), **config** (`midnight_config_get/_set/_unset`, `midnight_cache_clear`). The newest tools are the consequential ones: an agent can now drive the **full contract lifecycle** (deploy / call / read state), not just move tokens, and a **two-step confirmation** tool makes it *show the human what it is about to spend and get a yes before anything executes*. Point any MCP client at the binary, such as Claude Code (`.mcp.json`), Claude Desktop, Cursor, VS Code (Copilot), or Windsurf:

Reviewer: [BLOCKER] This is a 157-word paragraph. No developer will parse a wall of inline tool names.

Split the tool listing into a table (the same format used at line 21 on this page):

```
The **midnight-wallet-cli** ships an **MCP server** (`midnight-wallet-mcp`) with
31 tools that mirror the CLI commands, so an AI agent can operate a self-custody wallet.

| Group | MCP tools |
|---|---|
| Wallet and info | `midnight_wallet_generate`, `_list`, `_use`, `_info`, `_remove`, `midnight_info`, `midnight_address` |
| Transactions | `midnight_balance`, `_transfer`, `_airdrop`, `midnight_inspect_cost` |
| DUST | `midnight_dust_register`, `_status` |
| Local network | `midnight_localnet_up`, `_stop`, `_down`, `_status`, `_clean` |
| Config | `midnight_config_get`, `_set`, `_unset`, `midnight_cache_clear` |

The newest additions drive the **full contract lifecycle** (deploy, call, read state)
and include a **two-step confirmation** tool that shows the human what the agent is about
to spend before executing.
```

This is structural, so I have left it as a comment rather than a suggestion.

Author: Done — converted the 31-tool paragraph into the grouped table; the contract-lifecycle and two-step-confirmation notes now follow it. (04141105)
